### PR TITLE
Initial changes for Spark to Snowpark migration

### PR DESCRIPTION
**Pull request by navinrathore**
Only limited files included

### CODE REVIEW

**Review by sonalgoyal — CHANGES_REQUESTED** [review]
thanks! please make the following changes
a. No generic * imports - change to specific class, package imports
b. SaveMode - get from Snowflake
c. Instead of commenting out spark stuff in imports, just edit 
d. Why are jdk versions bumped up to 11? Is that what snowflake needs? Our Spark stuff is on jdk8, makes sense to have the same version unless Snowpark is incompatible.
e. Why do we need scala collection classes in TrainingDataFinder? I dont see them used anywhere?

**Comment by navinrathore** [comment]
Java 11 is only supported version.

**Comment by navinrathore** [comment]
A few places the snowpark api requires Seq object. One way is do it in below mentioned way. There could be alternatives such as directly create scala Seq in the java class. 
// Util.java #470-473
List<Column> cols = new ArrayList<Column>();
cols.add(dupesActual.col(colName + "temp"));
cols.add(dupesActual.col(colName));		 
dupesActual = dupesActual.withColumn(colName,
        functions.concat(JavaConverters.asScalaIteratorConverter(cols.iterator()).asScala().toSeq()));

File: client/src/main/java/zingg/client/util/Util.java [hunk 1]
```diff
@@ -462,14 +460,17 @@ public static String getCurrentLabelPath(String path) throws IOException {
 		
 	}
 	
-	public static Dataset<Row> addUniqueCol(Dataset<Row> dupesActual, String colName) {
+	public static DataFrame addUniqueCol(DataFrame dupesActual, String colName) {
 		String append = System.currentTimeMillis() + ":";
 		dupesActual = dupesActual.withColumn(colName + "temp", 
-				functions.lit(append));
+				com.snowflake.snowpark.functions.lit(append));
```
Comment (sonalgoyal) [inline client/src/main/java/zingg/client/util/Util.java:466]: why not let the preceding line work? The right import would fix it, no?

File: client/src/main/java/zingg/client/util/Util.java [hunk 2]
```diff
@@ -462,14 +460,17 @@ public static String getCurrentLabelPath(String path) throws IOException {
...
+				
+		List<Column> cols = new ArrayList<Column>();
+		cols.add(dupesActual.col(colName + "temp"));
+		cols.add(dupesActual.col(colName));		 
 		dupesActual = dupesActual.withColumn(colName,
-				functions.concat(dupesActual.col(colName + "temp"),
-						dupesActual.col(colName)));
-		dupesActual = dupesActual.drop(dupesActual.col(colName + "temp"));
+				functions.concat(JavaConverters.asScalaIteratorConverter(cols.iterator()).asScala().toSeq()));
```
Comment (sonalgoyal) [inline client/src/main/java/zingg/client/util/Util.java:472]: indentation?

Comment (sonalgoyal) [reply client/src/main/java/zingg/client/util/Util.java:472]: the logic is looking off to me? maybe we can discuss this.

File: core/src/main/java/zingg/TrainingDataFinder.java [hunk 3]
```diff
@@ -43,40 +27,41 @@ public TrainingDataFinder() {
...
-						
-						posPairs = trPairs.filter(trPairs.col(ColName.MATCH_FLAG_COL).equalTo(ColValues.MATCH_TYPE_MATCH));
-						negPairs = trPairs.filter(trPairs.col(ColName.MATCH_FLAG_COL).equalTo(ColValues.MATCH_TYPE_NOT_A_MATCH));
-						posPairs = posPairs.drop(ColName.MATCH_FLAG_COL, 
+					DataFrame trPairs = DSUtil.joinWithItself(trFile, ColName.CLUSTER_COLUMN, true);
+						posPairs = trPairs.filter(trPairs.col(ColName.MATCH_FLAG_COL).$eq$eq$eq(ColValues.MATCH_TYPE_MATCH));
+						posPairs = trPairs.filter(trPairs.col(ColName.MATCH_FLAG_COL).$eq$eq$eq(ColValues.MATCH_TYPE_MATCH));
+						posPairs = trPairs.filter(trPairs.col(ColName.MATCH_FLAG_COL).$eq$eq$eq(ColValues.MATCH_TYPE_MATCH));		
+						negPairs = trPairs.filter(trPairs.col(ColName.MATCH_FLAG_COL).$eq$eq$eq(ColValues.MATCH_TYPE_NOT_A_MATCH));
+						posPairs = posPairs.drop(new String[]{ColName.MATCH_FLAG_COL,
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/TrainingDataFinder.java:50]: Lines 46 and 48 and 47 and 49 look same to me?

File: core/src/main/java/zingg/TrainingDataFinder.java [hunk 4]
```diff
@@ -86,17 +71,17 @@ public void execute() throws ZinggClientException {
 						posPairs = posSamples;
 					}
 				}
-				posPairs = posPairs.cache();
-				if (negPairs!= null) negPairs = negPairs.cache();
+				// posPairs = posPairs.cache();
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/TrainingDataFinder.java:74]: why not cacheresult ?

File: core/src/main/java/zingg/TrainingDataFinder.java [hunk 5]
```diff
@@ -86,17 +71,17 @@ public void execute() throws ZinggClientException {
...
-				blocked = blocked.repartition(args.getNumPartitions(), blocked.col(ColName.HASH_COL)).cache();
-				Dataset<Row> blocks = DSUtil.joinWithItself(blocked, ColName.HASH_COL, true);
-				blocks = blocks.cache();	
+				DataFrame blocked = sample.map(new Block.BlockFunction(tree), RowEncoder.apply(Block.appendHashCol(sample.schema())));
+				//blocked = blocked.repartition(args.getNumPartitions(), blocked.col(ColName.HASH_COL)).cache();
+				DataFrame blocks = DSUtil.joinWithItself(blocked, ColName.HASH_COL, true);
+				blocks = blocks.cacheResult();	
 				//TODO HASH Partition
-				if (negPairs!= null) negPairs = negPairs.persist(StorageLevel.MEMORY_ONLY());
+				//if (negPairs!= null) negPairs = negPairs.persist(StorageLevel.MEMORY_ONLY());
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/TrainingDataFinder.java:84]: please use cacheResult

File: core/src/main/java/zingg/TrainingDataFinder.java [hunk 6]
```diff
@@ -131,52 +116,52 @@ public void execute() throws ZinggClientException {
...
 
-	public void writeUncertain(Dataset<Row> dupesActual) {
+	public void writeUncertain(DataFrame dupesActual) {
 		//input dupes are pairs
 		dupesActual = DFUtil.addClusterRowNumber(dupesActual, spark);
 		dupesActual = Util.addUniqueCol(dupesActual, ColName.CLUSTER_COLUMN );		
-		Dataset<Row> dupes1 = DSUtil.alignDupes(dupesActual, args);
-		Dataset<Row> dupes2 = dupes1.orderBy(ColName.CLUSTER_COLUMN);
+		DataFrame dupes1 = DSUtil.alignDupes(dupesActual, args);
+		DataFrame dupes2 = dupes1.sort(dupes1.col(ColName.CLUSTER_COLUMN), null);
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/TrainingDataFinder.java:124]: Is null necessary here?

File: core/src/main/java/zingg/TrainingDataFinder.java [hunk 7]
```diff
@@ -131,52 +116,52 @@ public void execute() throws ZinggClientException {
...
 		return PipeUtil.getTrainingDataUnmarkedPipe(args);
 	}
 
-	public Dataset<Row> getUncertain(Dataset<Row> dupes) {
+	public DataFrame getUncertain(DataFrame dupes) {
 		//take lowest positive score and highest negative score in the ones marked matches
-		Dataset<Row> pos = dupes.filter(dupes.col(ColName.PREDICTION_COL).equalTo(ColValues.IS_MATCH_PREDICTION));
-		pos = pos.sort(asc(ColName.SCORE_COL)).cache();
+		DataFrame pos = dupes.filter(dupes.col(ColName.PREDICTION_COL).$eq$eq$eq(ColValues.IS_MATCH_PREDICTION));
+		pos = pos.sort(pos.col(ColName.SCORE_COL), null);
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/TrainingDataFinder.java:137]: please check if you can call the method without null?

File: core/src/main/java/zingg/TrainingDataFinder.java [hunk 8]
```diff
@@ -131,52 +116,52 @@ public void execute() throws ZinggClientException {
...
 		if (LOG.isDebugEnabled()) {
 			LOG.debug("num pos " + pos.count());	
 		}
 		pos = pos.limit(10);
-		Dataset<Row> neg = dupes.filter(dupes.col(ColName.PREDICTION_COL).equalTo(ColValues.IS_NOT_A_MATCH_PREDICTION));
-		neg = neg.sort(desc(ColName.SCORE_COL)).cache();
+		DataFrame neg = dupes.filter(dupes.col(ColName.PREDICTION_COL).$eq$eq$eq(ColValues.IS_NOT_A_MATCH_PREDICTION));
+		neg = neg.sort(com.snowflake.snowpark.functions.col(ColName.SCORE_COL), null);
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/TrainingDataFinder.java:143]: why is an explicit com.snowflake...call needed here?

Comment (sonalgoyal) [reply core/src/main/java/zingg/TrainingDataFinder.java:143]: also null?

File: core/src/main/java/zingg/util/DSUtil.java [hunk 9]
```diff
@@ -77,22 +80,22 @@
...
-		Dataset<Row> lines1 = getPrefixedColumnsDS(lines); 
+    public static DataFrame joinWithItself(DataFrame lines, String joinColumn, boolean filter) throws Exception {
+		DataFrame lines1 = getPrefixedColumnsDS(lines); 
 		return join(lines, lines1, joinColumn, filter);
 	}
 	
-	public static Dataset<Row> joinWithItselfSourceSensitive(Dataset<Row> lines, String joinColumn, Arguments args) throws Exception {
-		Dataset<Row> lines1 = getPrefixedColumnsDS(lines).cache();
+	public static DataFrame joinWithItselfSourceSensitive(DataFrame lines, String joinColumn, Arguments args) throws Exception {
+		DataFrame lines1 = getPrefixedColumnsDS(lines); //.cache();
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/util/DSUtil.java:89]: cacheResult ?

File: core/src/main/java/zingg/util/DSUtil.java [hunk 10]
```diff
@@ -154,19 +162,20 @@
 		}*/
 		
 		
-		Dataset<Row> dupes2 = dupesActual.select(JavaConverters.asScalaIteratorConverter(cols1.iterator()).asScala().toSeq());
-	 	dupes2 = dupes2.toDF(dupes1.columns()).cache();
+		DataFrame dupes2 = dupesActual.select((Column [])cols1.toArray()); //.asScalaIteratorConverter(cols1.iterator()).asScala().toSeq());
+		//DataFrame dupes2 = dupesActual.select(JavaConverters.asScalaIteratorConverter(cols1.iterator()).asScala().toSeq());
+	 	dupes2 = dupes2.toDF(dupes1.schema().names()); //columns()).cache();
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/util/DSUtil.java:167]: cahceResult

File: core/src/main/java/zingg/util/DSUtil.java [hunk 11]
```diff
@@ -206,36 +215,38 @@
...
+	private static DataFrame getTraining(Session snow, Arguments args, Pipe p) {
+		DataFrame trFile = null;
 		try{
-			trFile = PipeUtil.read(spark, 
+			trFile = PipeUtil.read(snow, 
 					false, false, p); 
 			LOG.warn("Read marked training samples ");
-			trFile = trFile.drop(ColName.PREDICTION_COL);
-			trFile = trFile.drop(ColName.SCORE_COL);				
+			trFile = trFile.drop(ColName.PREDICTION_COL, null);
```
Comment (sonalgoyal) [inline core/src/main/java/zingg/util/DSUtil.java:239]: null - is it necessary?

**Review by sonalgoyal — CHANGES_REQUESTED** [review]

**Pull request closed without merging**